newIDE/app/public/External/Utils/pathEditor.js

```diff
@@ -0,0 +1,148 @@
+    initialResourcePath = initialResourcePath + '/NewFile' + extension;
+  }
+
+  const headerObject = {
+    saveOptions: {
+      folderPath: initialResourcePath.substring(
+        0,
+        initialResourcePath.lastIndexOf('/')
+      ),
+      name: initialResourcePath.substring(
+        initialResourcePath.lastIndexOf('/') + 1,
+        initialResourcePath.lastIndexOf('.')
+      ),
+      extension: initialResourcePath.substring(
+        initialResourcePath.lastIndexOf('.'),
+        initialResourcePath.length
+      ),
+      projectBasePath: projectPath,
+    }
+  };
+
+  // create the dom elements of the ui
+  headerObject.saveFolderLabel = editorContentDocument.createElement('label');
+  headerObject.saveFolderLabel.style = headerStyle.saveFolderLabel;
+  headerObject.saveFolderLabel.textContent = headerObject.saveOptions.folderPath;
+  parentElement.appendChild(headerObject.saveFolderLabel);
+
+  headerObject.nameInput = editorContentDocument.createElement('input');
+  headerObject.nameInput.type = 'text';
+  headerObject.nameInput.style = headerStyle.nameInput;
+  headerObject.nameInput.value = headerObject.saveOptions.name;
+  parentElement.appendChild(headerObject.nameInput);
+
+  headerObject.fileExistsLabel = editorContentDocument.createElement('label');
+  headerObject.fileExistsLabel.style = headerStyle.fileExistsLabel;
+  parentElement.appendChild(headerObject.fileExistsLabel);
+
+  headerObject.saveButton = editorContentDocument.createElement('button');
+  headerObject.saveButton.textContent = 'Save';
+  headerObject.saveButton.style = headerStyle.saveButton;
+  parentElement.appendChild(headerObject.saveButton);
+
+  headerObject.cancelButton = editorContentDocument.createElement('button');
+  headerObject.cancelButton.textContent = 'Cancel';
+  headerObject.cancelButton.style = headerStyle.cancelButton;
+  parentElement.appendChild(headerObject.cancelButton);
+  headerObject.cancelButton.addEventListener('click', cancelChangesFunction);
+
+  headerObject.setFolderButton = editorContentDocument.createElement('button');
+  headerObject.setFolderButton.textContent = 'Set Folder';
+  headerObject.setFolderButton.style = headerStyle.setFolderButton;
+  parentElement.appendChild(headerObject.setFolderButton);
+
+
+  // From here on we hook the dom with the imported or local methods via event listeners
+  headerObject.nameInput.addEventListener('input', () => {
+    renderPathEditor(headerObject)
+  });
+  headerObject.saveButton.addEventListener('click', () => {
+    saveToGDFunction(headerObject)
+  });
+  headerObject.saveFolderLabel.addEventListener('click', () => {
+    selectBaseFolderPath(headerObject)
+  });
+  headerObject.setFolderButton.addEventListener('click', () => {
+    selectBaseFolderPath(headerObject)
+  });
+  renderPathEditor(headerObject);
+  return headerObject //just in case it needs to be accessed from outside, we are returning it here
+}
+
+function renderPathEditor(headerObject) {
```

Comment:
renderXXX is a good function name.
Just for your interest:

Image that your function is called a "component"
Imagine that you call headerObject.saveOptions the "props"
Imagine that you can call other similar renderXXX functions.
Imagine that instead of manually mutating the DOM, you have a library that do it for you according to your render method.
Finally, imagine that you have a nice syntax to write your "render" method using a HTML like syntax `<like>this</like>`, but that is translated to JavaScript.

Picture it? That's React.js :)

Reply:
hahah yes, I was having a strange deja vu when refactoring it like that xD
Reactjs is fantastic and GD5 is an excellent example of why  😸 

I can rename it to just render if you like, it is now a purely internal function with no need to export anyway